src/mtbl_verify.c

```diff
@@ -0,0 +1,205 @@
+	const char *prefix,
+	const uint64_t file_offset,
+	const uint64_t bytes_data_blocks,
+	const uint64_t count_data_blocks)
+{
+	bool res = true;
+	uint8_t *data;
+	uint64_t offset = 0;
+	uint64_t bytes_consumed = 0;
+
+	/* Map the data blocks. */
+	data = mmap(NULL, bytes_data_blocks, PROT_READ, MAP_PRIVATE, fd, file_offset);
+	if (data == MAP_FAILED) {
+		fprintf(stderr, "%s: mmap() failed\n", prefix);
+		return false;
+	}
+
+	/* Verify each data block. */
+	for (uint64_t block = 0; block < count_data_blocks; block++) {
+		uint32_t block_crc, calc_crc;
+		uint32_t raw_contents_size;
+		uint8_t *raw_contents;
+
+		/* Get the data block size. */
+		raw_contents_size = mtbl_fixed_decode32(&data[offset + 0]);
+
+		/* Bounds check. */
+		bytes_consumed += 2 * sizeof(uint32_t);
+		bytes_consumed += raw_contents_size;
+		if (bytes_consumed > bytes_data_blocks) {
+			clear_line_stdout();
+			fprintf(stderr, "%s: Error: Block length (%u bytes) exceeds "
+				"total data length at "
+				"data block %" PRIu64 " (%" PRIu64 " bytes into file)\n",
+				prefix,
+				raw_contents_size,
+				block,
+				file_offset + offset);
+			return false;
+		}
+
+		/* CRC32C check. */
+		raw_contents = &data[offset + 2 * sizeof(uint32_t)];
+		block_crc = mtbl_fixed_decode32(&data[offset + 1 * sizeof(uint32_t)]);
+		calc_crc = mtbl_crc32c(raw_contents, raw_contents_size);
+		if (block_crc != calc_crc) {
+			clear_line_stdout();
+			fprintf(stderr, "%s: Error: block_crc (%08x) != calc_crc (%08x) at "
+				"data block %" PRIu64 " (%" PRIu64 " bytes into file)\n",
+				prefix,
+				block_crc,
+				calc_crc,
+				block,
+				file_offset + offset);
+			res = false;
+			break;
+		}
+
+		/* Progress report. */
+		if ((block % block_progress_interval) == 0) {
+			if (isatty(STDOUT_FILENO)) {
+				clear_line_stdout();
+				printf("%s: %'" PRIu64 " out of %'" PRIu64 " blocks (%.2f%%) OK",
+				       prefix,
+				       block,
+				       count_data_blocks,
+				       100.0 * (block + 0.0) / (count_data_blocks + 0.0));
+				fflush(stdout);
+			}
+		}
+
+		/* Update 'offset' to the next data block. */
+		offset += 2 * sizeof(uint32_t);
+		offset += raw_contents_size;
+	}
+
+	/* Unmap the data blocks. */
+	munmap(data, bytes_data_blocks);
+
+	clear_line_stdout();
+	return res;
+}
+
+static bool
+verify_file(const char *fname)
+{
+	bool res = true;
+	int fd;
+	struct mtbl_reader *r;
+	const struct mtbl_metadata *m;
+
+	fd = open(fname, O_RDONLY);
+	if (fd < 0) {
+		fprintf(stderr, "%s: open() failed: %s\n", fname, strerror(errno));
+		return false;
+	}
+
+	r = mtbl_reader_init_fd(fd, NULL);
+	if (r == NULL) {
+		close(fd);
+		fprintf(stderr, "%s: mtbl_reader_init_fd() failed\n", fname);
+		return false;
+	}
+
+	m = mtbl_reader_metadata(r);
+
+	uint64_t count_data_blocks = mtbl_metadata_count_data_blocks(m);
+	uint64_t bytes_data_blocks = mtbl_metadata_bytes_data_blocks(m);
+	uint64_t index_offset = mtbl_metadata_index_block_offset(m);
+
+	assert(index_offset >= bytes_data_blocks);
```

Comment:
`index_offset` is the absolute offset in the file where the index block is located, so it will always be at a larger offset in the file than the total byte size of the data blocks, in a properly encoded MTBL file. Probably this `assert()` should just be converted into an error return with a suitable failure message written to `stdout`.

However, I've discovered an embarrassing but unrelated bug in how `mmap()` is called from `verify_data_blocks()` which does indeed cause `mtbl_verify` to not work with files that have foreign prefixes.

Stay tuned, I'll rebase this branch on top of `next` and fix the `mmap()` issue.